src/linear_algebra/symplectic_group.lean

```diff
@@ -0,0 +1,242 @@
+end
+
+variables [fintype l]
+
+lemma J_squared : (J l R) ⬝ (J l R) = -1 :=
+begin
+  unfold J,
+  rw from_blocks_multiply,
+  simp only [matrix.zero_mul, matrix.neg_mul, zero_add, neg_zero', matrix.one_mul, add_zero],
+  rw [← neg_zero, ← matrix.from_blocks_neg,← from_blocks_one],
+end
+
+lemma J_inv : (J l R)⁻¹ = -(J l R) :=
+begin
+  refine matrix.inv_eq_right_inv _,
+  rw [matrix.mul_neg, J_squared],
+  exact neg_neg 1,
+end
+
+variables (R)
+variables [nontrivial R]
+
+lemma J_det_mul_J_det [rc : fact (ring_char R ≠ 2)] : (det (J l R)) * (det (J l R)) = 1 :=
+begin
+  rw [←det_mul, J_squared],
+  rw [←one_smul R (-1 : matrix _ _ R)],
+  rw [smul_neg, ←neg_smul, det_smul],
+  simp only [fintype.card_sum, det_one, mul_one],
+  rw neg_one_pow_eq_one_iff_even (ring.neg_one_ne_one_of_char_ne_two (fact.elim rc)),
+  exact even_add_self _
+end
+
+lemma J_det_unit [fact (ring_char R ≠ 2)] : is_unit (det (J l R)) :=
+is_unit_iff_exists_inv.mpr ⟨det (J l R), J_det_mul_J_det _ _⟩
+
+end matrix
+
+end J_matrix_lemmas
+
+namespace matrix
+
+variables (l) (R) [decidable_eq l] [fintype l] [comm_ring R]
+
+/-- The group of symplectic matrices over a ring `R`. -/
+def symplectic_group : submonoid (matrix (l ⊕ l) (l ⊕ l)  R) :=
+{ carrier := { A | A ⬝ (J l R) ⬝ Aᵀ = J l R},
+  mul_mem' :=
+  begin
+    intros a b ha hb,
+    simp only [mul_eq_mul, set.mem_set_of_eq, transpose_mul] at *,
+    rw [←matrix.mul_assoc, a.mul_assoc, a.mul_assoc, hb],
+    exact ha,
+  end,
+  one_mem' := by simp }
+
+end matrix
+
+namespace symplectic_group
+
+variables {l} {R} [decidable_eq l] [fintype l] [comm_ring R]
+
+open matrix
+
+lemma mem_iff {A : matrix (l ⊕ l) (l ⊕ l)  R} :
+  A ∈ symplectic_group l R ↔ A ⬝ (J l R) ⬝ Aᵀ = J l R :=
+by simp [symplectic_group]
+
+instance coe_matrix : has_coe (symplectic_group l R) (matrix (l ⊕ l) (l ⊕ l)  R) := ⟨subtype.val⟩
+
+section coe_lemmas
+
+variables (A B : symplectic_group l R)
+
+@[simp] lemma mul_val : ↑(A * B) = (A : matrix (l ⊕ l) (l ⊕ l) R) ⬝ (B : matrix (l ⊕ l) (l ⊕ l) R)
+:= rfl
+
+@[simp] lemma one_val : ↑(1 : symplectic_group l R) = (1 : matrix (l ⊕ l) (l ⊕ l)  R) := rfl
+
+lemma mul_mem {A B : matrix (l ⊕ l) (l ⊕ l) R}
+  (hA : A ∈ symplectic_group l R) (hB : B ∈ symplectic_group l R) :
+  A ⬝ B ∈ symplectic_group l R :=
+submonoid.mul_mem _ hA hB
+
+end coe_lemmas
+
+section symplectic_J
+
+variables (l) (R)
+
+lemma J_mem : (J l R) ∈ symplectic_group l R :=
+begin
+  rw mem_iff,
+  unfold J,
+  rw [from_blocks_multiply, from_blocks_transpose, from_blocks_multiply],
+  simp,
+end
+
+/-- The canonical skew-symmetric matrix as an element in the symplectic group. -/
+def sym_J : symplectic_group l R := ⟨J l R, J_mem l R⟩
+
+variables {l} {R}
+
+@[simp] lemma coe_J : ↑(sym_J l R) = J l R := rfl
+
+end symplectic_J
+
+variables {R} {A : matrix (l ⊕ l) (l ⊕ l) R}
+
+lemma neg_mem (h : A ∈ symplectic_group l R) : -A ∈ symplectic_group l R :=
+begin
+  rw mem_iff at h ⊢,
+  simp [h],
+end
+
+variables [nontrivial R]
```

Comment:
Again, can you please check if this is really needed? For example in `symplectic_det` it is not.